### PR TITLE
fix(navigator): rtl compute wind angle to select best land approach based on rally point location instead of home location

### PR DESCRIPTION
**Pull request by JonasPerolini**
As a first step for: https://github.com/PX4/PX4-Autopilot/pull/26973 (fyi @mrpollo). 

VTOL rally-point land approaches are defined around the rally-point land location, but RTL was choosing between them using the bearing from home. That works only when the landing point is effectively home. For remote safe points it can select the wrong loiter circle even when the available approach data is correct.

Behavior changes:
- compute the wind-alignment bearing from the rally-point land location instead of from home

```
const float wind_angle = wrap_pi(get_bearing_to_next_waypoint(_home_pos_sub.get().lat,
							 _home_pos_sub.get().lon, vtol_land_approaches.approaches[i].lat,
							 vtol_land_approaches.approaches[i].lon) - wind_direction);
```

becomes: 

```
const float wind_angle = wrap_pi(get_bearing_to_next_waypoint(vtol_land_approaches.land_location_lat_lon(0),
				 vtol_land_approaches.land_location_lat_lon(1), vtol_land_approaches.approaches[i].lat,
				 vtol_land_approaches.approaches[i].lon) - wind_direction);
```

Other changes:
- stop collecting approaches once `land_approaches_s::num_approaches_max` is reached
- Fix corner case in which the land approach could be chosen from a different rally point when rally points are close by
- Validate the associated rally point before exposing attached approaches
- Refactor code with smaller helpers to reduce computations
- Add a 500_ms timeout to loadWait calls to avoid dead locks
- Add unit test coverage

### CODE REVIEW

**Comment by mrpollo** [comment]
Unrelated to your PR, but I just wanted to drop a message to say our CI actions are now able to post comments on fork PRs!

File: src/modules/navigator/test/test_RTL.cpp [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include <gtest/gtest.h>
```
[the remark this answers is not in the document]
Comment (JonasPerolini) [reply src/modules/navigator/test/test_RTL.cpp:41]: comes from: 

```
if(BUILD_TESTING)
	add_subdirectory(test)
endif() 
```

[the remark this answers is not in the document]
Comment (JonasPerolini) [reply src/modules/navigator/test/test_RTL.cpp:41]: comes from:

```
if(BUILD_TESTING)
	add_subdirectory(test)
endif() 
```

**Comment by sfuhrer** [comment]
Looks reasonable high level, I would like to get @mbjd 's review on it.

File: src/modules/navigator/rtl.cpp [hunk 2]
```diff
@@ -378,18 +389,12 @@ void RTL::setRtlTypeAndDestination()
...
-
-			// set loiter position to destination initially, overwrite for VTOL if land approaches exist
-			landing_loiter.lat = destination.lat;
-			landing_loiter.lon = destination.lon;
-			landing_loiter.height_m = NAN;
-
-			if (_vehicle_status_sub.get().is_vtol
-			    && (_vehicle_status_sub.get().vehicle_type == vehicle_status_s::VEHICLE_TYPE_FIXED_WING)
-			    && rtl_land_approaches.isAnyApproachValid()) {
-				landing_loiter = chooseBestLandingApproach(rtl_land_approaches);
```
Comment (JonasPerolini) [inline src/modules/navigator/rtl.cpp:392]: Note that this check is now inside of `selectLandingApproach` to simplify `setRtlTypeAndDestination` and to avoid checking all approaches if we know we won't use them.

File: src/modules/navigator/rtl.cpp [hunk 3]
```diff
@@ -454,9 +459,12 @@ PositionYawSetpoint RTL::findClosestSafePoint(float min_dist, uint8_t &safe_poin
 				const float dist{get_distance_to_next_waypoint(_global_pos_sub.get().lat, _global_pos_sub.get().lon, mission_safe_point.lat, mission_safe_point.lon)};
 
 				PositionYawSetpoint safepoint_position;
-				setSafepointAsDestination(safepoint_position, mission_safe_point);
```
Comment (JonasPerolini) [inline src/modules/navigator/rtl.cpp:457]: Note that `setSafepointAsDestination` was replaced by `extractValidSafePointPosition` which returns false if the altitude frame is not valid or if the position is not valid. This change ensures we do not fly to an invalid approach. Especially important if we had other valid approaches.

File: src/modules/navigator/rtl.cpp [hunk 4]
```diff
@@ -454,9 +449,12 @@ PositionYawSetpoint RTL::findClosestSafePoint(float min_dist, uint8_t &safe_poin
...
 
-				const bool current_safe_point_has_approaches{hasVtolLandApproach(safepoint_position)};
```
Comment (JonasPerolini) [inline src/modules/navigator/rtl.cpp:459]: Note that instead of calling `hasVtolLandApproach(safepoint_position)` we now call `hasVtolLandApproach(current_seq, _home_pos_sub.get().alt)`. The important change here is the `current_seq`. 

The old code would go through all land approahces starting from index zero (which is not efficient) meaning that if two rally points (A, B) are close enough, we could have ended up with the location from rally point B but the approach from rally point A.

File: src/modules/navigator/rtl.cpp [hunk 5]
```diff
@@ -693,94 +704,181 @@ bool RTL::reverseIsFurther() const
...
 
-	for (int current_seq = 0; current_seq < _stats.num_items; ++current_seq) {
-		mission_item_s mission_item{};
+	if (vtol_land_approaches.isAnyApproachValid()) {
+		landing_approach = chooseBestLandingApproach(vtol_land_approaches);
+	}
 
-		bool success_mission_item = _dataman_cache_safepoint.loadWait(static_cast<dm_item_t>(_stats.dataman_id), current_seq,
-					    reinterpret_cast<uint8_t *>(&mission_item),
-					    sizeof(mission_item_s));
```
Comment (JonasPerolini) [inline src/modules/navigator/rtl.cpp:751]: Note that the `loadWait` call was updated to include the standard `500_ms` timeout

File: src/modules/navigator/test/test_RTL.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrpollo) [reply src/modules/navigator/test/test_RTL.cpp:41]: is this a false positive?

[the remark this answers is not in the document]
Comment (JonasPerolini) [reply src/modules/navigator/test/test_RTL.cpp:41]: Yes @mrpollo, I added a comment here earlier https://github.com/PX4/PX4-Autopilot/pull/27019 it's due to 

```
if(BUILD_TESTING)
	add_subdirectory(test)
endif()  
```

**Comment by JonasPerolini** [comment]
The PR is ready to be reviewed @mbjd (please note that the PR is not very large, 960 lines come from the unit tests)

Summary of the main changes: 

1. Compute the wind-alignment bearing from the rally-point land location instead of from home
2. Fix corner case in which the land approach could be chosen from a different rally point when rally points are close by  
3. Validate the associated rally point before exposing attached approaches
4. stop collecting approaches once land_approaches_s::num_approaches_max is reached
5. Refactor code with smaller helpers to reduce computations 
6. Add a `500_ms` timeout to `loadWait` calls to avoid dead locks
7. Add unit test coverage

Tested only using unit tests: `make tests TESTFILTER=test_RTL`

**Comment by mrpollo** [comment]
Sorry didnt mean to close the PR, I was trying to prevent the damn bot from spamming you.

**Review by hamishwillee — COMMENTED** [review]
Does this need docs?

Minimally a release note to https://docs.px4.io/main/en/releases/main, but it might be useful to add something to Return mode page(s) too.

@mbjd

File: src/modules/navigator/rtl.cpp [hunk 6]
```diff
@@ -552,31 +555,40 @@ void RTL::setLandPosAsDestination(PositionYawSetpoint &rtl_position, mission_ite
...
-		break;
+		return false;
 	}
+
+	position.lat = safe_point_item.lat;
+	position.lon = safe_point_item.lon;
+	position.yaw = NAN;
+
+	return PX4_ISFINITE(position.lat) && PX4_ISFINITE(position.lon) && PX4_ISFINITE(position.alt)
+	       && !((fabs(position.lat) < NULL_ISLAND_THRESHOLD_DEG) && (fabs(position.lon) < NULL_ISLAND_THRESHOLD_DEG))
```
Comment (mbjd) [inline src/modules/navigator/rtl.cpp]: What exactly does this null island check protect against, is there a case where we have lat/lon of exactly zero which needs to be considered invalid? As far as I can see this was not checked before. Could you add a comment explaining why this is done (or remove if unnecessary)? 

Also, if exact zero is the invalid case, I would prefer to check for `fabsf(x) < FLT_EPSILON`, to be as close as possible to float equality without making build warnings. (after all the rule about never checking float equality is only really necessary for results of calculations in which errors can build up...)

File: src/modules/navigator/rtl.cpp [hunk 7]
```diff
@@ -552,31 +555,40 @@ void RTL::setLandPosAsDestination(PositionYawSetpoint &rtl_position, mission_ite
...
-		rtl_position.lat = mission_safe_point.lat;
-		rtl_position.lon = mission_safe_point.lon;
-		rtl_position.alt = mission_safe_point.altitude + _home_pos_sub.get().alt; // alt of safe point is rel to home
+	case NAV_FRAME_GLOBAL_RELATIVE_ALT:
+	case NAV_FRAME_GLOBAL_RELATIVE_ALT_INT:
+		if (!PX4_ISFINITE(home_altitude_amsl)) {
+			return false;
+		}
+
+		position.alt = safe_point_item.altitude + home_altitude_amsl; // alt of safe point is rel to home
```
Comment (mbjd) [inline src/modules/navigator/rtl.cpp:578]: This is not very related to your changes but more a general question - does an altitude relative to home make sense if rally points can be arbitrarily far away? I guess we have NAV_MIN_GND_DIST to protect against the worst but it could still be confusing for users

File: src/modules/navigator/rtl.cpp [hunk 8]
```diff
@@ -552,31 +555,40 @@ void RTL::setLandPosAsDestination(PositionYawSetpoint &rtl_position, mission_ite
...
 		break;
 
 	default:
-		mavlink_log_critical(_navigator->get_mavlink_log_pub(), "RTL: unsupported MAV_FRAME\t");
-		events::send<uint8_t>(events::ID("rtl_unsupported_mav_frame"), events::Log::Error, "RTL: unsupported MAV_FRAME ({1})",
-				      mission_safe_point.frame);
-		break;
+		return false;
```
Comment (mbjd) [inline src/modules/navigator/rtl.cpp:582]: Double checking: Is the removal of this error intentional? As far as I can see in the caller we fall back to just not using the rally point

**Review by mbjd — COMMENTED** [review]
Thanks @JonasPerolini! 

I tested it with `sihsim_standard_vtol` (merged main again for #27075, simulate wind with #26467) and it seems to fix the issue! Also the thorough unit tests are very appreciated :) 

I cannot say I follow every last detail of your refactor but looks good. Added some minor nits / questions.

**Comment by mbjd** [comment]
(messed up merge commit, corrected in force-push above)

File: src/modules/navigator/rtl.cpp (hunk 7 above)
Comment (JonasPerolini) [reply src/modules/navigator/rtl.cpp:578]: I agree (and I do not use this frame), but getting rid of it will probably be hard (and should not be hidden in this PR).

File: src/modules/navigator/rtl.cpp (hunk 8 above)
Comment (JonasPerolini) [reply src/modules/navigator/rtl.cpp:582]: The removal was intentional to avoid spamming this message. To be consistent we would need one message per error so two new messages

```
	if (safe_point_item.nav_cmd != NAV_CMD_RALLY_POINT) {
		return false;
	}
```

```
		if (!PX4_ISFINITE(home_altitude_amsl)) {
			return false;
		}
```

What do you think?

File: src/modules/navigator/rtl.cpp (hunk 6 above)
Comment (JonasPerolini) [reply src/modules/navigator/rtl.cpp]: The aim is to prevent a fly away. lat,lon can be zero if `struct mission_item_s` is not init or has missing values. Since `mission_item_s mission_safe_point;` is loaded from the cache I perfered to be safe and include a null check. 

I've updated the code here: https://github.com/PX4/PX4-Autopilot/pull/27004/changes/2b9d24fb32df136d32e4907e96ef61d6cdf5f678 to use `fabsf(x) < FLT_EPSILON`

File: src/modules/navigator/rtl.cpp (hunk 8 above)
Comment (mbjd) [reply src/modules/navigator/rtl.cpp:582]: Makes complete sense as long as the caller has sensible fallback, I agree with not outputting a message every time. 

Maybe though we can warn if there were points but none of them were valid? But no strong opinion from my side, current state is also good

File: src/modules/navigator/rtl.cpp (hunk 6 above)
Comment (mbjd) [reply src/modules/navigator/rtl.cpp]: Double checking now I see that FLT_EPSILON is actually bigger than the initial `1e-7` so that part of my comment was misguided... also I did not mean specifically float as opposed to double, I just did not catch it was double - but no way for you to know that. Sorry if I was unclear.

What do you think about e2bb0c1608930ed8f28d787a71924cd933ee4fbe? (dangling commit - cherrypick, modify or ignore)
 - add explaining comment for readability
 - decrease size of island incorrectly flagged invalid

**Comment by JonasPerolini** [comment]
@mbjd thanks for the review. Please note that now that https://github.com/PX4/PX4-Autopilot/pull/26993 is merged, I've moved `VectorMissionItemStore` into a new file to avoid duplications: https://github.com/PX4/PX4-Autopilot/pull/27004/changes/d860078386f8ae788d52672a1d982833b8714bab. (The commit also renames `makePositionFromOffset` to `makePositionYawSetpointFromOffset`)

File: src/modules/navigator/rtl.cpp (hunk 6 above)
Comment (JonasPerolini) [reply src/modules/navigator/rtl.cpp]: Thanks!

Comment (JonasPerolini) [reply src/modules/navigator/rtl.cpp]: @mbjd I pushed something before seeing your commit. I've re-applied your changes here https://github.com/PX4/PX4-Autopilot/pull/27004/changes/c41abc69bd7d9574fb9bc53ab06f3c31933fcc96. Closing thread

File: src/modules/navigator/test/test_RTL.cpp [hunk 9]
```diff
@@ -0,0 +1,937 @@
...
+
+#include <cmath>
+#include <string>
+#include <vector>
+
+#include <drivers/drv_hrt.h>
+#include <lib/geo/geo.h>
+#include <parameters/param.h>
+#include <uORB/uORB.h>
+#include <uORB/topics/home_position.h>
```
[the remark this answers is not in the document]
Comment (JonasPerolini) [reply src/modules/navigator/test/test_RTL.cpp]: fixed here https://github.com/PX4/PX4-Autopilot/pull/27004/changes/55f0eaddb3ad7205032e1e2d15612b470fc9f4a7

[the remark this answers is not in the document]
Comment (JonasPerolini) [reply src/modules/navigator/test/test_RTL.cpp]: fixed here https://github.com/PX4/PX4-Autopilot/pull/27004/changes/55f0eaddb3ad7205032e1e2d15612b470fc9f4a7

**Review by mbjd — DISMISSED** [review]

File: src/modules/navigator/test/test_RTL.cpp [hunk 10]
```diff
@@ -0,0 +1,936 @@
...
+
+#include <cmath>
+#include <string>
+#include <vector>
+
+#include <drivers/drv_hrt.h>
+#include <lib/geo/geo.h>
+#include <parameters/param.h>
+#include <uORB/uORB.h>
+#include <uORB/topics/vehicle_status.h>
```
[the remark this answers is not in the document]
Comment (JonasPerolini) [reply src/modules/navigator/test/test_RTL.cpp:51]: solved here: https://github.com/PX4/PX4-Autopilot/pull/27004/changes/9bcd3fd82aa8a3a032f9f7904a0bb8ab654cbfe9

File: src/modules/navigator/rtl.cpp (hunk 8 above)
Comment (JonasPerolini) [reply src/modules/navigator/rtl.cpp:582]: @mbjd I can include the warning but note that it would require a new class variable e.g. `_invalid_rally_points_warning_reported` to only report the warning once. Flag to be reset here: `if (_opaque_id != _stats.opaque_id)`

**Comment by JonasPerolini** [comment]
Hey @mbjd after you've approved the changes I pushed two suggestions from the git bot that I now reverted. This dismissed your approval. Sorry about that.

---

> Does this need docs?
> 
> Minimally a release note to https://docs.px4.io/main/en/releases/main, but it might be useful to add something to Return mode page(s) too.

Hey @hamishwillee I've included a docs change here: https://github.com/PX4/PX4-Autopilot/pull/27004/changes/1c1903e09d2a59a7cbdda690ec759f476ced1d55

Please note that I also added a line for https://github.com/PX4/PX4-Autopilot/pull/26993 because it was merged without change log entry.

File: docs/en/releases/main.md [hunk 11]
```diff
@@ -34,6 +34,8 @@ Please continue reading for [upgrade instructions](#upgrade-guide).
 
 ## Other changes
 
+- RTL_MISSION_FAST and RTL_MISSION_FAST_REVERSE now skip DO_JUMP commands (loops). ([PX4-Autopilot#26993: fix(navigator): goToNextPositionItem skip loops when required](https://github.com/PX4/PX4-Autopilot/pull/26993))
```
Comment (hamishwillee) [inline docs/en/releases/main.md:37]: Is RTL_MISSION_FAST (and so on) an actual identifier? If you mean RTL_TYPE then link to those please.

File: docs/en/flight_modes/return.md [hunk 12]
```diff
@@ -92,7 +92,8 @@ In this return type the vehicle:
 - If the destination is a mission landing pattern it will follow the pattern to land.
 - If the destination is a rally point or home it will [land or wait](#loiter-landing-at-destination) at descent altitude (depending on landing parameters).
   By default an MC or VTOL in MC mode will land, and a fixed-wing vehicle circles at the descent altitude.
-  A VTOL in FW mode aligns its heading to the destination point, transitions to MC mode, and then lands.
+  A VTOL in FW mode can first fly a VTOL approach loiter associated with that destination; if several are defined, PX4 chooses the most wind-aligned one.
+  It then aligns its heading to the destination point, transitions to MC mode, and lands.
```
Comment (hamishwillee) [inline docs/en/flight_modes/return.md:96]: This is cool. Just to be clear, if you had 2 rally points and all the other conditions said you had to land at one, it would pick the one that ws best to land at based on the wind?

If so, that's what this says so no change needed.

File: docs/en/flight_modes/return.md [hunk 13]
```diff
@@ -220,6 +221,7 @@ The RTL parameters are listed in [Parameter Reference > Return Mode](../advanced
...
+| <a id="RTL_APPR_FORCE"></a>[RTL_APPR_FORCE](../advanced_config/parameter_reference.md#RTL_APPR_FORCE)       | [VTOL FW only] If set, home or rally-point RTL destinations are only considered when a valid VTOL approach loiter is defined for that landing location. Mission landing patterns are unaffected.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
```
Comment (hamishwillee) [inline docs/en/flight_modes/return.md:224]: "[VTOL FW only] If set, home or rally-point RTL destinations are only considered when a valid VTOL approach loiter is defined for that landing location." - what does the term "valid VTOL approach loiter is define for ..." mean? 

Note, this might be provided by other context in this page - I'm only seeing the changes right here.

File: docs/en/releases/main.md (hunk 11 above)
Comment (JonasPerolini) [reply docs/en/releases/main.md]: Good catch, updated to: 

Fast mission Return modes ([RTL_TYPE](../advanced_config/parameter_reference.md#RTL_TYPE) = 2 and 4) now skip `DO_JUMP` commands (loops) while following the mission path. ([PX4-Autopilot#26993: fix(navigator): goToNextPositionItem skip loops when required](https://github.com/PX4/PX4-Autopilot/pull/26993))

File: docs/en/flight_modes/return.md (hunk 12 above)
Comment (JonasPerolini) [reply docs/en/flight_modes/return.md]: The RP is chosen by distance. Once the rally point is chosen, wind is the tie breaker if that specific rally point has several VTOL approaches. I changed "if several are defined" → "if several approach loiters are defined for the destination" to make that explicit.

File: docs/en/flight_modes/return.md (hunk 13 above)
Comment (JonasPerolini) [reply docs/en/flight_modes/return.md:228]: This param already existed: https://docs.px4.io/v1.15/en/advanced_config/parameter_reference#return-to-land

> RTL force approach landing
> Comment: Only consider RTL point, if it has an approach defined.

Updated the param description: 

```
Only consider RTL point, if it has a valid (finite position/altitude) VTOL approach loiter.
A VTOL approach loiter is a loiter circle attached to the rally point or home
that the vehicle uses to descend before transitioning.
```

I've added explanations in both the generic and VTOL return pages

_A _VTOL approach loiter_ is a loiter circle attached to the rally point or home that the vehicle uses to descend before transitioning._

**Comment by JonasPerolini** [comment]
Hey @hamishwillee I've updated the docs to address your comments (see latest commit) and squashed all the other commits to clean the history.

**Comment by JonasPerolini** [comment]
Tested again in SIM by adding debug prints. No changes were made (except from docs) since @mbjd approved the PR so the PR is ready to be merged if the docs is ok @hamishwillee 

----
Test 1: 3 land approaches, at 100, 75, and 50 meters 
- `candidate idx=0 bearing-wind=37.3 deg `
- `candidate idx=1 bearing-wind=71.4 deg `
- `candidate idx=2 bearing-wind=127.3 deg`

The one with the smallest bearing-wind angle was chosen: `RTL VTOL approach: chose candidate idx=0 abs angle=37.3 deg`, and the drone stopped spiraling at the correct altitude.

Note that I also tested with angles of `-48 deg` and `-14 deg` and the one with `-14` was chosen (i.e. we properly check the abs value).

---
Test 2: See image below. 
- The rally point has 2 approaches
   - 100m AGL, north from RP (0 deg)
   - 50m AGL, west from RP (-90 deg)
- Home is south of the west approach.

Results: 

- `choosing with wind dir 1.0 deg`
- `candidate idx=0 bearing-wind=-1.0 deg`
- `candidate idx=1 bearing-wind=-89.0 deg`
- `chose candidate idx=0 abs angle=1.0 deg`

The drone properly flew to the north approach. If the angle had been computed from home, the west approach (which is north from home) would have had 1 deg bearing-wind angle and the UAV would have chosen the west approach.
 
rp

Traj output: 

Traj

---
Test 3: repeated test 2 but with `-45 deg` wind. The UAV properly flew to the west approach, descended to 50m and then flew to the RP.

Results: 

- `choosing with wind dir -90.0 deg`
- `candidate idx=0 bearing-wind=90.0`
- `candidate idx=1 bearing-wind=1.9 deg`
- `chose candidate idx=1 abs angle=1.9 deg`

File: docs/en/releases/1.18.md [hunk 14]
```diff
@@ -120,7 +120,7 @@ Please continue reading for [upgrade instructions](#upgrade-guide).
 
 ### VTOL
 
-- TBD
+- [Return mode VTOL Rally Point Approach Loiters](../flight_modes_vtol/return.md#vtol-rally-point-approach-loiters): VTOLs returning in fixed-wing mode can now use the approach loiter associated with the selected home/rally landing location (instead of home) to choose the most wind-aligned valid approach (if several are defined). ([PX4-Autopilot#27004: fix(navigator): rtl compute wind angle to select best land approach based on rally point location instead of home location](https://github.com/PX4/PX4-Autopilot/pull/27004))
```
Comment (hamishwillee) [inline docs/en/releases/1.18.md:123]: FYI moved from main to here, because we should get this in today.

File: docs/en/flight_modes_vtol/return.md [hunk 15]
```diff
@@ -47,13 +47,47 @@ If returning as a fixed-wing, the vehicle:
...
-  - Transition to MC mode and land.
+  - Fly from the approach loiter to the return destination (the rally point or home location).
+  - Transition to MC mode at the destination and land.
 
     Note that [NAV_FORCE_VT](../advanced_config/parameter_reference.md#NAV_FORCE_VT) is ignored: the vehicle will always land as a multicopter for these destinations.
 
+#### VTOL Rally Point Approach Loiters
+
+VTOL _rally point approach loiters_ are [MAV_CMD_NAV_LOITER_TO_ALT](https://mavlink.io/en/messages/common.html#MAV_CMD_NAV_LOITER_TO_ALT) items associated with a particular rally point ([MAV_CMD_NAV_RALLY_POINT](https://mavlink.io/en/messages/common.html#MAV_CMD_NAV_RALLY_POINT)) in a [Rally plan](../flying/plan_safety_points.md).
+They define options for where the VTOL can descend to the rally point approach altitude, and how it will approach the rally point.
```
Comment (hamishwillee) [inline docs/en/flight_modes_vtol/return.md:65]: Do you set the heading and xtrack fields? I imagine you specify that it should head to the next point (the rally point) following the direct xtrack (or does it try for the centre xtrack?). 

Below I added a tip that this is not defined in the MAVLink specification. I think this feature is cool but the fact we've just gone and done it has a few issues
- We're not compliant with the standard because that states only rally points can be in the plan.
- Getting it into the standard is harder.
- We're adding code to block invalid items. Others will not expect this so no one can write compliant code.
- QGC typically argues against non-standard code, so getting this in the plan becomes harder. Do you have a PR for that?

I'll try get it in the spec https://github.com/mavlink/mavlink-devguide/pull/712

FYI only I've split this section out and removed a little of the repetition. It is linked from above. Having it a details section looks good in that it is _right there_. But this is really odd and different behaviour. I'm going to need a very clear link to it, so a new section is better.

**Review by hamishwillee — APPROVED** [review]
@JonasPerolini There is possible tweaking to be done, but I think we should merge this today to get it into v1.18. 

Are you OK with that and to do any updates as a post process.

**Comment by hamishwillee** [comment]
@JonasPerolini I have merged because it would be ideal to get this into v1.18 and it already passed and had tech approval. BUT I would still like to clarify the docs as discussed in https://github.com/PX4/PX4-Autopilot/pull/27004#pullrequestreview-4423819243 - it matters to users if they need to set the xtrack and heading values, and if so, to what. If not, to what.

FWIW This PR is a pretty cool feature. Next time you do something that adds a new supported type to a rally plan, please let me know first  :-)

**Comment by hamishwillee** [comment]
@JonasPerolini @mrpollo FYI the implementation of Rally points in this PR is considered non standard by MAVLink. The implication of this is that it isn't likely to be adopted by other flight stacks and ground stations.

The discussion is in the chain here between me and Peter Barker: https://github.com/mavlink/mavlink-devguide/pull/712#issuecomment-4618770329

Peter's argument is that 
- The spec explicitly states that rally points may only have rally point items.
- Rally points are not designed for this purpose - this effectively makes them a mini mission. 
- He points to prior art in the mission plan that does things "like this" - specifically using something like [MAV_CMD_DO_RETURN_PATH_START](https://mavlink.io/en/messages/common.html#MAV_CMD_DO_RETURN_PATH_START) in a mission to define alternative return path landing points.

I am not in complete agreement on all that. Mostly because I don't like what we do with embedding non-mission behaviour into missions, such as these return paths. I much prefer the logic that if you're defining behaviour that can be used outside of a mission it is defined outside of a mission. On the other hand, I wouldn't just change rally points without consultation.

My preference would be that we try collaboratively agree a way to do this. My preference is that this would be a new plan type that more generically does what you have done here - i.e. including Fixed wing landings. But if not, I'd rather do this in missions with agreement of all stakeholders, than in a rally plan.

**Comment by JonasPerolini** [comment]
Thanks for the heads up @hamishwillee, I've added a comment there. Please note that this PR did not introduce the VTOL land approaches, they were merged here https://github.com/PX4/PX4-Autopilot/pull/21049 3 years ago,

This PR just fixed how the selection is done, adds unit tests, and documents it.
